# Host COI data selection strategy

Because the primers are known to amplify host COI we developed a database of potential bat sequences representing species known to occupy the Cypress Creek National Wildlife Refuge. In addition, we expanded our search to include additional bat species from neighboring states such as Kentucky, Indiana, Tennesee, Arkansas, and Missouri. Finally, other bat and bird species reference sequences were collected. These species were known to have guano being processed in the same lab that this study was conducted and we wanted to be able to detect potential instances of cross contamination. We prioritized selecting representative sequences for each species by identifying published articles over undocumented Genbank or Barcode of Life Database (BOLD) submissions, and selected the full length (either ~ 500 or 650 bp) COI Sanger sequence to short reads where possible.

# Host data sources
Full references are available at the end of this document. Citations and links included where possible. Abbreviated species names are explained in the table following the attributed papers (see below).

## Bat references

1. (Clare et al., 2011) [supplementary Table S1](https://doi.org/10.1371/journal.pone.0022648.s003).
  - bat species included: CACA, CAPE, CASU, DERO, EPBR, EPFU, EUMA, GLSO, LABL, MOMO, MYNI, MYRI, NOAL, NOLE, PHDI, PHHA, PHST, PTGY, PTPA, PTPE, RHPU, RHNA, RHTU, STLI, TOSA, TRCI

2. (Vonhof, Russell, and Miller-Butterworth, 2015) [PopSet 726974368](https://www.ncbi.nlm.nih.gov/popset/?term=726974368)
  - bat species included: MYSO, MYSE, and MYLE

3. (Streicker et al., 2010) [PopSet 301344216](https://www.ncbi.nlm.nih.gov/popset/?term=301344216)
  - bat species included: MYAU, MYLU, PESU, EPFU, NYHU, LABO, COTO, LACI, LANO

4. (Walker et al. 2016) [supplementary Table S2](https://doi.org/10.1371/journal.pone.0162342.s002)
  - bat species included: EUGL, EUUN, RHBI, STPA

5. (Korstian, Williams, and Hale, 2016) [supplementary table 12686_2015_509_MOESM1_ESM.txt](https://static-content.springer.com/esm/art%3A10.1007%2Fs12686-015-0509-4/MediaObjects/12686_2015_509_MOESM1_ESM.txt)
  - bat species included: LAEG

6. from BOLD (direct submission)[http://v4.boldsystems.org/index.php/Public_RecordView?processid=BPBAT246-14]
  - bat species included: MOSI

## Bird references

Additional host COI reference sequences for a few bird species were added because samples processed in the same lab as this study.

1. (Kerr et al., 2007)
  - bird species included: [ABTO](https://www.ncbi.nlm.nih.gov/nuccore/DQ433094), [ATFL](https://www.ncbi.nlm.nih.gov/nuccore/DQ433033), [LUWA](https://www.ncbi.nlm.nih.gov/nuccore/DQ433253)
2. (Kerr et al., 2009)
  - bird species included: [CORF](https://www.ncbi.nlm.nih.gov/nuccore/GQ481508)
3. (Kerr, 2011) [supplementary Table S2](https://onlinelibrary.wiley.com/action/downloadSupplement?doi=10.1111%2Fj.1755-0998.2011.03049.x&file=MEN_3049_sm_TableS2.pdf)
  - bird species included: MAWA, JAWE, SOSP, WCSP, WIWA, YEWA
4. (Hebert et al., 2004)
  - bird species included: BEVI
5. (Schindel et al., 2011) [BioProject PRJNA81595](https://www.ncbi.nlm.nih.gov/bioproject/?term=PRJNA81595)
  - bird species included: [BEWR](https://www.ncbi.nlm.nih.gov/nuccore/JQ176472), [YBCH](https://www.ncbi.nlm.nih.gov/nuccore/HM033416)
6. (Saitoh et al., 2015)
  - bird species included: [JABW](https://www.ncbi.nlm.nih.gov/nuccore/AB843412)
7.  from BOLD (direct submission)
  - bird species included: [BHCO](https://www.ncbi.nlm.nih.gov/nuccore/HM033586), [BHGR](http://v4.boldsystems.org/index.php/Public_RecordView?processid=BOM2215-11), [BLPH](https://www.ncbi.nlm.nih.gov/nuccore/DQ434059), [CLSW](https://www.ncbi.nlm.nih.gov/nuccore/AY666213), [COYE](https://www.ncbi.nlm.nih.gov/nuccore/HM033476), [HAAM](https://www.ncbi.nlm.nih.gov/nuccore/NC_025603.1?report=genbank), [LEGO](http://v4.boldsystems.org/index.php/Public_RecordView?processid=BOM852-07), [MELT](http://v4.boldsystems.org/index.php/Public_RecordView?processid=BVIET015-08), [RBLE](http://v4.boldsystems.org/index.php/Public_RecordView?processid=SIBHI082-11), VERD

## Species codes
The following table identifies the species name for each of the abbreviated bat and bird species codes used above.

| code | taxa name |
| --- | --- |
|ABTO|Melozone aberti|
|ATFL|Myiarchus cinerascens|
|BEVI|Vireo bellii|
|BEWR|Thryomanes bewickii|
|BHCO|Molothrus ater|
|BHGR|Pheucticus melanocephalus|
|BLPH|Sayornis nigricans|
|CACA|Carollia castanea|
|CAPE|Carollia perspicillata|
|CASU|Carollia subrufa|
|CLSW|Petrochelidon pyrrhonota|
|CORF|Carpodacus erythrinus|
|COYE|Geothlypis trichas|
|DERO|Desmodus rotundus|
|EPBR|Eptesicus brasiliensis|
|EPFU|Eptesicus fuscus|
|EPFUR|Eptesicus furinalis|
|EUGL|Eumops glaucinus|
|EUMA|Euderma maculatum|
|EUUN|Eumops underwoodi|
|GLSO|Glossophaga soricina|
|HAAM|Hemignathus virens|
|JABW|Horornis diphone|
|JAWE|Zosterops japonicus|
|LABL|Lasiurus blossevillii|
|LAEG|Lasiurus ega|
|LEGO|Spinus psaltria|
|LUWA|Vermivora luciae|
|MACR|_Carpodacus erythrinus_ used as substitute|
|MAWA|Setophaga magnolia|
|MELT|Garrulax canorus|
|MOMO|Molossus molossus|
|MOSI|Molossus sinaloae|
|MYNI|Myotis nigricans|
|MYRI|Myotis riparius|
|NOAL|Noctilio albiventris|
|NOLE|Noctilio leporinus|
|PHDI|Phyllostomus discolor|
|PHHA|Phyllostomus hastatus|
|PHST|Phylloderma stenops|
|PTGY|Pteronotus gymnonotus|
|PTPA|Pteronotus parnellii|
|PTPE|Pteronotus personatus|
|RBLE|Leiothrix lutea|
|RHBI|Rhogeessa bickhami|
|RHNA|Rhynchonycteris naso|
|RHTU|Rhogeessa tumida|
|SOSP|Melospiza melodia|
|STLI|Sturnira lilium|
|STPA|Sturnira parvidens|
|TOSA|Tonatia saurophila|
|TRCI|Trachops cirrhosus|
|VERD|Auriparus flaviceps|
|WCSP|Zonotrichia leucophrys|
|WIWA|Cardellina pusilla|
|YBCH|Icteria virens|
|YEWA|Setophaga petechia|

# QIIME formatting of host sequences
The above fasta sequences were saved as a temporary fasta file `tmp.fa` and converted into QIIME-formatted `.qza` taxonomy and sequence files for import using the following script (original concatenated fasta file starts as `tmp2.fa`)
```
seqtk seq -l0 tmp.fa > tmp2.fa
cat tmp2.fa | paste - - | awk -F "\t" '{print $2}' > tmp.seqs.txt
cat tmp2.fa | paste - - | awk -F "\t" '{print $1}' > tmp.header.txt
sed 's/>//g' tmp.header.txt | cut -f 1 -d ';' | awk '{ printf("%s_%s\n",$0,i++);next;} { print $0;}' > tax.accessions.txt
cut -d ';' -f 1 --complement tmp.header.txt | sed 's/tax=//g' > tmp.taxstring.txt


R
library(data.table)
setwd('/mnt/lustre/macmaneslab/devon/guano/BOLDdb')
df <- fread(file="tmp.taxstring.txt", header = FALSE, sep = ",", fill=TRUE)
df <- data.frame(lapply(df, function(x) {gsub("*.\\:", "", x)}))
df$V1 <- paste("k__", df$V1, sep = "")
df$V2 <- paste("p__", df$V2, sep = "")
df$V3 <- paste("c__", df$V3, sep = "")
df$V4 <- paste("o__", df$V4, sep = "")
df$V5 <- paste("f__", df$V5, sep = "")
df$V6 <- paste("g__", df$V6, sep = "")
df$V7 <- paste("s__", df$V7, sep = "")
df2 <- data.frame(paste(df$V1, df$V2, df$V3, df$V4, df$V5, df$V6, df$V7, sep = ";"))
write.table(df2, file = "/mnt/lustre/macmaneslab/devon/guano/BOLDdb/taxstring.txt", quote=FALSE, row.names = FALSE, col.names = FALSE)
q()
n

paste tax.accessions.txt taxstring.txt > qiime.host.txt
sed 's/^/>/g' tax.accessions.txt > tmp.tax.accessions.txt
paste tmp.tax.accessions.txt tmp.seqs.txt | tr '\t' '\n' > qiime.host.fa

rm tmp.seqs.txt tmp.header.txt tax.accessions.txt tmp.taxstring.txt tmp.tax.accessions.txt taxstring.txt
```

Files were then imported into QIIME-format with the following pair of functions:
```
qiime tools import \
  --type 'FeatureData[Sequence]' \
  --input-path qiime.host.fa \
  --output-path host_seqs.qza

qiime tools import \
  --type 'FeatureData[Taxonomy]' \
  --input-format HeaderlessTSVTaxonomyFormat \
  --input-path qiime.host.txt \
  --output-path host_taxonomy.qza
```

The text and QIIME-formatted files mentioned above are available in the [Open Science Frameworks repo](https://osf.io/qju3w/) for this project. (DOI 10.17605/OSF.IO/QJU3W)

# References

Clare, E. L., Lim, B. K., Fenton, M. B., & Hebert, P. D. N. (2011). Neotropical Bats: Estimating Species Diversity with DNA Barcodes. PLoS ONE, 6(7), e22648. doi:10.1371/journal.pone.0022648Â 

Hebert,P.D., Stoeckle,M.Y., Zemlak,T.S. and Francis,C.M. (2004) Identification of Birds through DNA Barcodes. PLoS Biol. 2 (10), E312.

Kerr,K.C., Stoeckle,M.Y., Dove,C.J., Weigt,L.A., Francis,C.M. and Hebert,P.D. (2007) Comprehensive DNA barcode coverage of North American birds. Mol. Ecol. Notes 7(4), 535-543.

Kerr KC, Birks SM, Kalyakin MV, Red'kin YA, Koblik EA, Hebert PD. (2009) Filling the gap - COI barcode resolution in eastern Palearctic birds. Front Zool. Dec 9;6:29. doi: 10.1186/1742-9994-6-29.

Kerr,K.C. (2011) Searching for evidence of selection in avian DNA barcodes. Mol Ecol Resour 11 (6), 1045-1055.

Korstian JM, AM Hale, VJ Bennett, and DA Williams. 2016. Using DNA barcoding to improve bat carcass identification at wind farms in the United States. Conservation Genetics Resources 8:27-34.

Saitoh,T., Sugita,N., Someya,S., Iwami,Y., Kobayashi,S., Kamigaichi,H., Higuchi,A., Asai,S., Yamamoto,Y. and Nishiumi,I. (2015) DNA barcoding reveals 24 distinct lineages as cryptic bird species candidates in and around the Japanese Archipelago. Mol Ecol Resour 15 (1), 177-186.

Schindel DE, Stoeckle MY, Milensky C, Trizna M, Schmidt B, Gebhard C, Graves G. (2011) DNA barcodes of bird species in the national museum of natural history, smithsonian institution, USA.", Zookeys, Dec 8;(152):87-92.

Streicker DG, Turmelle AS, Vonhof MJ, Kuzmin IV, McCracken GF, Rupprecht CE. Host phylogeny constrains cross-species emergence and establishment of rabies virus in bats. Science. 2010 Aug 6;329(5992):676-9. doi: 10.1126/science.1188836.

Vonhof,M.J., Russell,A.L. and Miller-Butterworth,C.M. (07-08-2015) Range-Wide Genetic Analysis of Little Brown Bat (Myotis lucifugus) Populations: Estimating the Risk of Spread of White-Nose Syndrome. PLoS ONE 10:(7)E0128713

Walker FM, Williamson CHD, Sanchez DE, Sobek CJ, Chambers CL (2016) Species From Feces: Order-Wide Identification of Chiroptera From Guano and Other Non-Invasive Genetic Samples. PLOS ONE 11(9): e0162342. https://doi.org/10.1371/journal.pone.0162342
